### PR TITLE
[#1097] normalize jars by inode

### PR DESCRIPTION
**Pull request by eubnara**
Thank you for submitting a pull request! 

Please make sure you have signed our [Contributor License Agreement (CLA)][].  
We are not asking you to assign copyright to us, but to give us the right to distribute your code without restriction. We ask this of all contributors in order to assure our users of the origin and continuing existence of the code.  
You only need to sign the CLA once.

- [x] I have signed the [Contributor License Agreement (CLA)][]

[Contributor License Agreement (CLA)]: https://www.elastic.co/contributor-agreement/

Hello, I request this pull request to solve #1097 issue.
I tried to normalize duplicate jars by getting and comparing inodes of jars.
I think it is more reasonable to investigate the inodes of a jar than to just remove `.` or `..` in urls.

### CODE REVIEW

File: mr/src/itest/java/org/elasticsearch/hadoop/VersionTest.java [hunk 1]
```diff
@@ -37,16 +37,15 @@ public void testNormalize() throws Exception {
         List<URL> urls = new ArrayList<URL>();
         urls.add(new URL("jar:file:/tmp/mesos/slaves/3014350f-cd05-44af-9c9c-3974bdeed86e-S1/frameworks/3014350f-cd05-44af-9c9c-3974bdeed86e-0016/executors/3014350f-cd05-44af-9c9c-3974bdeed86e-S1/runs/3697bc14-f0bd-48b9-8599-9683867eec63/elasticsearch-spark_2.10-2.2.0-m1.jar!/"));
         urls.add(new URL("jar:file:/tmp/mesos/slaves/3014350f-cd05-44af-9c9c-3974bdeed86e-S1/frameworks/3014350f-cd05-44af-9c9c-3974bdeed86e-0016/executors/3014350f-cd05-44af-9c9c-3974bdeed86e-S1/runs/3697bc14-f0bd-48b9-8599-9683867eec63/./elasticsearch-spark_2.10-2.2.0-m1.jar!/"));
+        File file = new File(urls[0].toString);
```
Comment (jbaiera) [inline mr/src/itest/java/org/elasticsearch/hadoop/VersionTest.java]: This doesn't compile for me? Did you mean `urls.get(0).toString()`?

File: mr/src/main/java/org/elasticsearch/hadoop/util/StringUtils.java [hunk 2]
```diff
@@ -409,57 +415,25 @@ public static IpAndPort parseIpAddress(String httpAddr) {
...
-                else {
-                    pathTokens.add(0, element);
-                }
-            }
+        HashMap<String, String> inoToFile = new HashMap<String, String>();
+        List<String> normalized = new ArrayList<String>();
+        for (URL url: urls) {
+            Path path = Paths.get(url.toString());
+            BasicFileAttributes attr = null;
+            attr = Files.readAttributes(path, BasicFileAttributes.class);
```
Comment (jbaiera) [inline mr/src/main/java/org/elasticsearch/hadoop/util/StringUtils.java]: My understanding is that these API's are java 7 APIs. Right now we're still supporting java 6 in the project. This isn't a hard stop, I'd like to deprecate and remove 6 and 7 support now that the older hadoop versions are removed, but that would have to happen on master before these API's can be used.

File: mr/src/main/java/org/elasticsearch/hadoop/util/StringUtils.java [hunk 3]
```diff
@@ -409,57 +415,25 @@ public static IpAndPort parseIpAddress(String httpAddr) {
...
+            Object fileKey = attr.fileKey();
+            String s = fileKey.toString();
+            String inode = s.substring(s.indexOf("ino=") + 4, s.indexOf(")"));
```
Comment (jbaiera) [inline mr/src/main/java/org/elasticsearch/hadoop/util/StringUtils.java]: Does this work for NTFS filesystems?

**Comment by eubnara** [comment]
Thanks for reviewing.
I didn't consider NTFS filesystem. It doesn't work on NTFS filesystem.
I try to go advance. Thank you.

**Comment by jbaiera** [comment]
Refreshing CI...

**Comment by jbaiera** [comment]
@eubnara Sorry for the delayed responses. I'd like to merge this since it's a great idea, but it looks like there are compilation errors in the code.

**Comment by jbaiera** [comment]
@eubnara just checking in about whether you were planning on making any further updates to this PR?

**Comment by jbaiera** [comment]
This PR has been open for almost a year and has had no updates on it. I'm going to go ahead and close it. I've got a separate PR going up soon that should resolve the issues that this one was attempting to fix.

**Pull request closed without merging**